Implement marshaler selection based on metadata model

jkoritzinsky opened the pull request:
Create a basic data model to represent parsed attribute metadata either with the old system or with the new attribute model described in the StructMarshalling design doc. (Reading the metadata into the model for the new attributes TBD).

Update MarshalerGenerators.TryCreate to use the new data model to determine generators and have more specific coverage of primitive types.

Make the bool marshaler more compatible with the new data model (use a parameterized marshaler instance instead of reading the metadata in the marshaler).

Implement more efficient blittable byref marshalling so we can more efficiently marshal blittable structs (which are coming with the new model)

Implement delegate->function pointer marshaller generator to show the flexibility of the pattern matching approach for the MarshalerGenerators.TryCreate method.

I've validated that the Roslyn code-gen for the pattern matching switch is as optimal as doing a nested switch statement model more similar to what we currently do in the runtime. I will say that the readability for the more complex patterns could use some improvement, so I'm open to suggestions/recommendations.

AaronRobinsonMSFT commented on DllImportGenerator/DllImportGenerator/StubCodeContext.cs:
```diff
@@ -46,6 +46,10 @@ public enum Stage
 
         public Stage CurrentStage { get; protected set; }
 
+        public abstract bool PinningSupported { get; }
```
I think I just need someone to elaborate on the inheritance design aspect of this.

jkoritzinsky replied:
The idea I had was that the `StubCodeContext` would be the base type for all of the generators (struct, reverse P/Invoke, P/Invoke, etc) and the subclasses would implement the actual generation. So the P/Invoke generator (the only one we have today) would set both of these to true, but a struct generator would set both to false.

AaronRobinsonMSFT replied:
That model is surprising when the `StubCodeContext` has a function like `GetIdentifiers()` or `GenerateReturnNativeIdentifier()` which doesn't seem to have a relation of a struct marshaller. I am okay with inheritance but not "just because". Can't the context be passed in? I don't see the obvious win for this architecture and it is starting to crack given the above methods on it - including the constants which also don't seem related to struct marshalling. I could be missing the bigger picture though.

jkoritzinsky replied:
`GetIdentifiers` can still make sense in the struct world, but I agree that `GenerateReturnNativeIdentifier` doesn't make sense. I think that `GenerateReturnNativeIdentifier` shouldn't be called from marshallers directly. It should be called from the generator or the behavior should be enabled by default.

jkoritzinsky replied:
Then the marshallers would only call `GetIdentifiers` to know which identifier names to use to refer to the managed and native values. A struct code generator could override the implementation of `GetIdentifiers` to return identifiers that point to fields on the managed and native types.

jkoritzinsky replied:
I've tried out a refactoring to make the StubCodeContext not have any knowledge of "return identifiers" in the base class. All of that logic now lives in the P/Invoke subclass.

jkoritzinsky commented on DllImportGenerator/DllImportGenerator/LanguageSupport.cs:
```diff
@@ -0,0 +1,10 @@
+namespace System.Runtime.CompilerServices
+{
+    // Define IsExternalInit type to support records.
+    internal class IsExternalInit
```
The `IsExternalInit` type is needed to use the new "init properties" feature in C# 9, which records use under the hood.

AaronRobinsonMSFT replied:
So our source generator needs to define this type so that Roslyn will enable the feature? Is this point in time or a common approach. Any documentation would be helpful here.

jkoritzinsky replied:
Yes. This is a common approach that has been used since .NET Framework 3.5 to enable features that don't require runtime support on downlevel runtimes (enabling LINQ query syntax on older .NET Framework versions is the first case I know of, our BCL IAsyncEnumerable package for .NET Standard 2.0 to enable `await foreach` is another example).

AaronRobinsonMSFT replied:
I know about the BCL side of the house. I am more intrigued by an extension point though. This seems like it is/should be a point in time issue and not something we will need to have long term or does Roslyn use this model throughout its extensibility architecture?

jkoritzinsky replied:
Since the source generator has to target .NET Standard 2.0, we'll have to keep this long term. If we were able to have the source generator target .NET 5.0, then we'd be able to remove it.

jkoritzinsky replied:
I'm not sure why Roslyn chose to not emit the type when it isn't present for IsExternalInit since for older features they used to emit the extra types needed for metadata when they weren't present. Maybe it's because this is a modreq instead of an attribute?

jkotas replied:
The official support matrix for C# versions and runtime versions is https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/configure-language-version#defaults . Records are C# 9 feature and so they are not officially supported for .net standard.

If you know what you are doing, you can make some of C# 9 features work using hacks like this.

AaronRobinsonMSFT replied:
>If you know what you are doing, you can make some of C# 9 features work using hacks like this.

I think that is the surprise for me. Adding a type to enable a language feature in the code being used is a bit indirect and abstruse. Not a fan, but get it.

jkotas replied:
The language feature is enabled by specifying C# 9 on the compiler command line (`<LangVersion>` in msbuild). This attribute is not enabling the language feature. It is "making the language feature work" in officially unsupported configuration.

jkoritzinsky commented on DllImportGenerator/DllImportGenerator/Marshalling/DelegateMarshaler.cs:
```diff
@@ -0,0 +1,115 @@
+                    {
+                        yield return ExpressionStatement(
+                            InvocationExpression(
+                                ParseName("global::System.GC.KeepAlive"),
```
We could definitely add a "keep alive" stage or something like that. This is used in the runtime marshalling by delegates and a few of the legacy handle types (HandleRef and CriticalHandle) and runtime handles (which are only useful in QCall scenarios)

AaronRobinsonMSFT replied:
I am fine with a "KeepAlive" stage. I only mention this because we are greenfield here and can learn from the past. I want to ensure we name things for clarity. Should we change "Clean" to "PostInvoke"? That is also an option in my opinion.

@elinor-fung Thoughts?

jkoritzinsky replied:
I think we should keep the "Clean" name because we want to reuse the stage names for other generators that don't have an "invoke" concept such as struct marshalling, which still needs the "clean" concept (although it will also put the keep-alive in the same function as clean to simplify codegen).

jkoritzinsky commented on DllImportGenerator/DllImportGenerator/Marshalling/MarshallingGenerator.cs:
```diff
@@ -65,26 +68,62 @@ public static bool TryCreate(TypePositionInfo info, out IMarshallingGenerator ge
-                case SpecialType.System_Double:
-                    generator = MarshallingGenerators.Numeric;
+                case { ManagedType: { SpecialType: SpecialType.System_SByte }, MarshallingAttributeInfo: null or MarshalAsInfo { UnmanagedType: 0 or UnmanagedType.I1 } }
+                    or { ManagedType: { SpecialType: SpecialType.System_Byte }, MarshallingAttributeInfo: null or MarshalAsInfo { UnmanagedType: 0 or UnmanagedType.U1 } }
```
We can emit an error diagnostic when `UnmanagedType` is `0` and simplify the cases here.

I'll set up the code to be ready to emit the diagnostic and add a TODO for that (we haven't plumbed through the diagnostic emit infra from the source generator down into the marshallers yet).

jkoritzinsky commented on DllImportGenerator/DllImportGenerator/TypePositionInfo.cs:
```diff
@@ -53,63 +39,67 @@ private TypePositionInfo()
         }
 
-        private static void UpdateWithAttributeData(IEnumerable<AttributeData> attributes, ref TypePositionInfo typeInfo)
+#nullable enable
```
I'd like to enable it everywhere, but I wasn't sure if we wanted to do that as part of this PR.

AaronRobinsonMSFT replied:
I don't really like the nullable thing at all, but if we are going to use it let's just enable it. It can wait for another PR though.

elinor-fung commented on DllImportGenerator/DllImportGenerator/Marshalling/BoolMarshaller.cs:
```diff
@@ -102,4 +100,28 @@ public IEnumerable<StatementSyntax> Generate(TypePositionInfo info, StubCodeCont
+    internal class VariantBoolMarshaller : BoolMarshallerBase
+    {
+        public VariantBoolMarshaller()
+            : base(PredefinedType(Token(SyntaxKind.ShortKeyword)), -1, 0)
```
https://docs.microsoft.com/openspecs/windows_protocols/ms-oaut/7b39eb24-9d39-498a-bcd8-75c38e5823d0?

AaronRobinsonMSFT replied:
Careful. Yes those are the correct values, but that is the automation transport layer specification not the actual API. We can debate whether that is okay to use or not, but technically that isn't required to be used in source.

jkoritzinsky commented:
I've added the metadata reader for the new attributes and added some tests around the BlittableType attribute.

I did do some work around non-blittable structs, but I've pulled that out of this PR and I'll put out a new PR in the future once we get to supporting them. From this work though, I found a bug where we were treating `in` parameters as by-value, and another bug in the `bool` unmarshalling code, which are both fixed in this PR.

Can I get another review pass?

AaronRobinsonMSFT commented on DllImportGenerator/DllImportGenerator/ManualTypeMarshallingHelper.cs:
```diff
@@ -0,0 +1,65 @@
+
+        public static IMethodSymbol? FindGetPinnableReference(ITypeSymbol type)
+        {
+            return type.GetMembers("GetPinnableReference")
```
Let's document this `GetPinnableReference()` method discovery. Why and what it means - https://docs.microsoft.com/dotnet/csharp/language-reference/proposals/csharp-7.3/pattern-based-fixed.

AaronRobinsonMSFT commented on DllImportGenerator/DllImportGenerator/ManualTypeMarshallingHelper.cs:
```diff
@@ -0,0 +1,65 @@
+#nullable enable
+
+using System.Linq;
```
Let's keep our Linq statements to a minimum. I appreciate they are convenient but they are also a huge perf sink.

jkoritzinsky replied:
I've tried to keep them pretty limited to just a little bit of member lookup. I might make some more helper methods to limit usage.

AaronRobinsonMSFT commented on DllImportGenerator/DllImportGenerator/ManualTypeMarshallingHelper.cs:
```diff
@@ -0,0 +1,65 @@
+        {
+            return nativeType.GetMembers(ToManagedMethodName)
+                    .OfType<IMethodSymbol>()
+                    .Any(m => m.Parameters.IsEmpty &&
```
@elinor-fung Long ago had an argument I liked for placing the logic operator on the left side. I do find it clearer. Thoughts?

```CSharp
.Any(m => m.Parameters.IsEmpty
    && !m.ReturnsByRef
    && !m.ReturnsByRefReadonly
    && SymbolEqualityComparer.Default.Equals(m.ReturnType, managedType)
    && !m.IsStatic);
```

jkoritzinsky replied:
I'm down for either. I didn't know if we had a pattern, so I just picked one. I'll switch them over in my next iteration.

AaronRobinsonMSFT commented on DllImportGenerator/DllImportGenerator/Marshalling/BoolMarshaller.cs:
```diff
@@ -8,24 +8,22 @@
 
-            return PredefinedType(Token(syntax));
+        public TypeSyntax AsNativeType(TypePositionInfo info)
+        {
```
Do we want to assert or validate the input here?

AaronRobinsonMSFT commented on DllImportGenerator/DllImportGenerator/Marshalling/BoolMarshaller.cs:
```diff
@@ -57,17 +55,14 @@ public IEnumerable<StatementSyntax> Generate(TypePositionInfo info, StubCodeCont
                     break;
                 case StubCodeContext.Stage.Marshal:
-                    // <nativeIdentifier> = (<nativeType>)(<managedIdentifier> ? 1 : 0);
+                    // <nativeIdentifier> = (<nativeType>)(<managedIdentifier> ? _trueValue : _falseValue);
```
Thank you! I really appreciate these small code snippets that distill the Roslyn code. There should really be one for every complex statement using the Roslyn API.

AaronRobinsonMSFT commented on DllImportGenerator/DllImportGenerator/StubCodeGenerator.cs:
```diff
@@ -16,20 +17,76 @@ private StubCodeGenerator(Stage stage)
+            }
+            else
+            {
+                return base.GetIdentifiers(info);
```
Let's assert and state the intuition for why the not managed and not native return condition.

jkoritzinsky commented on DllImportGenerator/DllImportGenerator/ManualTypeMarshallingAnalyzer.cs:
```diff
@@ -301,20 +301,17 @@ private void AnalyzeNativeMarshalerType(SymbolAnalysisContext context, ITypeSymb
                         continue;
                     }
-                    if (ctor.Parameters.Length == 1 && SymbolEqualityComparer.Default.Equals(type, ctor.Parameters[0].Type))
+                    if (ManualTypeMarshallingHelper.IsManagedToNativeConstructor(ctor, type))
```
Yes, I'll do that now that the condition is much simpler to read.
